oeps/oep-0008.rst

```diff
@@ -0,0 +1,399 @@
++-------------------+--------------------------------------------------------------+
+| quality           | check coding style                                           |
++-------------------+--------------------------------------------------------------+
+| test              | run all unit, integration, and acceptance tests for the repo |
++-------------------+--------------------------------------------------------------+
+| validate          | run tests and quality checks                                 |
++-------------------+--------------------------------------------------------------+
+| e2e (optional)    | run End to End tests against a deployed environment to       |
+|                   | verify the system as a user, initiating their actions        |
+|                   | through the UI                                               |
++-------------------+--------------------------------------------------------------+
+
+.. _Travis CI: https://travis-ci.org/
+.. _PyPI: https://pypi.python.org/pypi
+.. _Paver: https://pythonhosted.org/Paver/
+.. _Gulp: http://gulpjs.com/
+
+
+Python Language Specific
+------------------------
+
+Requirements
+^^^^^^^^^^^^
+
+* Core requirements are listed in the ``install_requires`` section of
+  ``setup.py``, and only modified when code changes are made that require a
+  new package or drop support for a previously supported one.
+
+* Requirements are managed via `pip-tools`_ and organized by which ones need
+  to be installed for different Python virtual environments.
+
+* Requirements files are kept either at root level or within a directory named
+  ``requirements``.
+
+* Direct dependencies are listed in ``requirements/*.in``.
+
+* Full pinned dependencies are auto-generated via ``make pip-compile`` and
+  saved in ``requirements/*.txt``. This file is tracked under version control.
+
+* For a developer to update package versions, she would run ``pip-compile
+  --upgrade`` and check in the changes to the .txt file(s).
+
+* Requirements are installed via ``make requirements`` which behind the scenes
+  will execute ``pip-sync``. Note that ``pip-sync`` updates a virtual env to
+  reflect exactly what is in the requirements files. This will
+  install/upgrade/uninstall everything necessary to match their contents.
+
+* Requirements that a developer also wants to use in their local development
+  environment (e.g. pudb) should be listed in a non-version-controlled
+  requirements file named ``private.txt``. This filename is included in the
+  ``pip-sync`` command and so its listed dependencies will be installed.
+
+* Requirements filename examples:
+
+  * ``requirements.*``: contains the base requirements
+
+  * ``test-requirements.*``: contains the requirements needed for writing and
+    running tests
+
+* Editable requirements format example:
+  ``-e git+https://github.com/edx/django-oauth2-provider.git@0.2.7-fork-edx-6a#egg =django-oauth2-provider==0.2.7-fork-edx-6``
+
```

Comment:
We may want to call out an explicit recommendation for the version numbers in the `egg=` stanza for repositories urls.

If you use `0.0`, pip will always try and reinstall the package, so you'll be guaranteed to get updates when they happen. Otherwise, you have to rely the version number of the git tag and the version number of the egg updating in tandem.